### PR TITLE
*: Refine LimitInputStream

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #5900 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/DataStreams/LimitTransformAction.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+#include <DataStreams/LimitTransformAction.h>
+#include <common/likely.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int ILLEGAL_TYPE_OF_COLUMN_FOR_FILTER;
+extern const int LOGICAL_ERROR;
+} // namespace ErrorCodes
```
Comment (SeaRise) [inline dbms/src/DataStreams/LimitTransformAction.cpp:24]:
```suggestion
```

File: dbms/src/DataStreams/LimitBlockInputStream.cpp [hunk 2]
```diff
@@ -24,67 +24,31 @@ LimitBlockInputStream::LimitBlockInputStream(
     size_t limit_,
     size_t offset_,
     const String & req_id,
-    bool always_read_till_end_)
-    : limit(limit_)
-    , offset(offset_)
-    , always_read_till_end(always_read_till_end_)
-    , log(Logger::get(req_id))
+    bool /*always_read_till_end_*/)
```
Comment (SeaRise) [inline dbms/src/DataStreams/LimitBlockInputStream.cpp:27]: It looks like `always_read_till_end_` is useless? Can we just delete it

File: dbms/src/DataStreams/LimitTransformAction.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+size_t LimitTransformAction::getLimit() const
+{
+    return limit;
+}
+
+bool LimitTransformAction::transform(Block & block)
+{
+    if (unlikely(!block))
+        return true;
+
```
Comment (SeaRise) [inline dbms/src/DataStreams/LimitTransformAction.cpp:50]: It seems that the related logic of offset has not been migrated?
```
        res = children.back()->read();
        if (!res)
            return res;
        rows = res.rows();
        pos += rows;
    } while (pos <= offset);
```

Comment (SeaRise) [reply dbms/src/DataStreams/LimitTransformAction.cpp:50]: will we remove offset in this pr?

File: dbms/src/DataStreams/LimitTransformAction.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    pos += rows;
+    if (pos >= rows && pos <= limit)
+    {
+        // give away the whole block
+        return true;
+    }
+    else
+    {
+        // give away a piece of the block
+        size_t start = std::max(
```
Comment (ywqzzy) [inline dbms/src/DataStreams/LimitTransformAction.cpp:59]: I will remove the cast later

File: dbms/src/DataStreams/LimitTransformAction.cpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    /// pos - how many lines were read, including the last read block
+    if (pos >= limit)
+    {
+        return false;
+    }
+
+    auto rows = block.rows();
+    pos += rows;
+    if (pos >= rows && pos <= limit)
```
Comment (SeaRise) [inline dbms/src/DataStreams/LimitTransformAction.cpp:51]: It seems that `pos >= rows` must be true

Comment (ywqzzy) [inline dbms/src/DataStreams/LimitTransformAction.cpp:51]:
```suggestion
    if (pos <= limit)
```

File: dbms/src/DataStreams/LimitBlockInputStream.cpp [hunk 6]
```diff
@@ -22,69 +22,31 @@ namespace DB
...
-    const String & req_id,
-    bool always_read_till_end_)
-    : limit(limit_)
-    , offset(offset_)
-    , always_read_till_end(always_read_till_end_)
-    , log(Logger::get(req_id))
+    const String & req_id)
+    : log(Logger::get(req_id))
+    , limit_transform_action(input->getHeader(), limit_)
+
```
Comment (SeaRise) [inline dbms/src/DataStreams/LimitBlockInputStream.cpp:28]:
```suggestion
```

**Review by SeaRise — APPROVED** [review]
lgtm

File: dbms/src/DataStreams/LimitTransformAction.h [hunk 7]
```diff
@@ -0,0 +1,39 @@
+
+
+
```
Comment (ywqzzy) [inline dbms/src/DataStreams/LimitTransformAction.h:3]:
```suggestion
```

**Pull request merged**